### DISCUSSION TITLE
Bug? - ORM Cascading objects into session - changed in 2.0

### DISCUSSION THREAD

**Discussion by cc-jj** [sqlalchemy/sqlalchemy#10307] [Usage Questions] [answered]
It seems the behavior for cascading objects into a session has changed in 2.0 - and I'm not sure this is documented anywhere. So perhaps this is a bug.

The documentation is here:
https://docs.sqlalchemy.org/en/20/tutorial/orm_related_objects.html#cascading-objects-into-the-session

### Steps to produce:
1. Create a parent record and add it to the session
2. Create a child record associated to the parent - do not add it to the session
3. Problem:
    - In version 1.4 the child is cascaded into the session
    - In version 2.0 the child is not

### Demonstration

<details><summary>ORM models</summary>

```python
from sqlalchemy.orm import DeclarativeBase, Session, relationship, Mapped
from sqlalchemy import Column, create_engine, Integer, ForeignKey


class Base(DeclarativeBase):
    id = Column(Integer, primary_key=True)


class Parent(Base):
    __tablename__ = "parent"
    children: Mapped[list["Child"]] = relationship(back_populates="parent")


class Child(Base):
    __tablename__ = "child"
    parent_id: Mapped[int] = Column(ForeignKey("parent.id"))
    parent: Mapped["Parent"] = relationship(back_populates="children")
```

</details>

<details><summary>Bug?</summary>

```python
engine = create_engine("sqlite:///:memory:", echo=True)

Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

with Session(engine) as session:
    parent = Parent()
    session.add(parent)
    
    child1 = Child(parent=parent)
    child2 = Child()
    parent.children.append(child2)
    
    assert parent in session  # succeeds
    assert child1 in session  # fails
    assert child2 in session  # fails

```

</details>

<details><summary>Workaround</summary>
<br>
If the parent is added to the session after the children are associated, then the children are also added to the session.

```python
with Session(engine) as session:
    parent = Parent()
    
    child1 = Child(parent=parent)
    child2 = Child()
    parent.children.append(child2)
    
    session.add(parent)

    assert parent in session  # succeeds
    assert child1 in session  # succeeds
    assert child2 in session  # succeeds

```

</details>

### Change in behavior from 1.4

I think this is a bug in 2.0 because it's an undocumented (as far as I can tell) change from 1.4.

<details><summary>Full 1.4 code</summary>

```python
from sqlalchemy.orm import declarative_base, Session, relationship, Mapped
from sqlalchemy import Column, create_engine, Integer, ForeignKey

class Base:
    id = Column(Integer, primary_key=True)


Base = declarative_base(cls=Base)

class Parent(Base):
    __tablename__ = "parent"
    children = relationship("Child", back_populates="parent")


class Child(Base):
    __tablename__ = "child"
    parent_id = Column(ForeignKey("parent.id"), nullable=False)
    parent = relationship("Parent", back_populates="children")


engine = create_engine("sqlite:///:memory:", echo=True)

Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

with Session(engine) as session:

    parent = Parent()
    session.add(parent)

    child1 = Child(parent=parent)
    child2 = Child()
    parent.children.append(child2)

    assert parent in session  # succeeds
    assert child1 in session  # succeeds
    assert child2 in session  # succeeds
```

</details>

**Comment by CaselIT** [comment, marked as answer]
Hi,

That's expected. It's documented here https://docs.sqlalchemy.org/en/20/changelog/migration_14.html#cascade-backrefs-behavior-deprecated-for-removal-in-2-0

Version 1.4 will also raise a warning when the 20 warning are enabled https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#migration-to-2-0-step-two-turn-on-removedin20warnings
